File: src/main/domain/tax/Taxation.java
package domain.tax;

//===========================================================//
//===========================================================//

public abstract class Taxation {
    //===========================================================//
    //===========================================================//
    // Public Field(s)

    static public final Taxation HUNGARY = new HungaryTaxation();

    //===========================================================//
    //===========================================================//
    // Constructor(s)

    protected Taxation() {}  // To prevent instantiation, but allow subclasses.
}

File: src/main/kotlin/domain/tax/Taxation.kt
package domain.tax

@Deprecated("WARNING : Taxation has threading problems and implementation hiccups, thus it is not representative of truly real world taxation especially over the long run.")
object Taxation {
    fun create(type: Type): ITaxation {
        return when (type) {
            is Type.Hungary -> HungaryTaxation()
        }
    }

    sealed interface Type {
        data object Hungary : Type

        companion object{
            val entries: List<Type> = listOf(
                Hungary
            )
        }
    }
}